packages/record/src/watchers/dom-watcher.ts

```diff
@@ -0,0 +1,218 @@
+
+      const parentVn = addedVNodesMap.get(parentId)
+
+      const isParentSVG = parentVn && parentVn.extra.isSVG
+
+      let vn = addedSiblingMap.get(node)!
+
+      if (isParentSVG && isVNode(vn)) {
+          ;(vn as VNode).extra.isSVG = true
+      }
+
+      addedNodes.push({
+          parentId,
+          nextId: nodeStore.getNodeId(node.nextSibling!) || null,
+          node: vn
+      })
+
+      if (isVNode(vn)) {
+          addedVNodesMap.set(vn.id, vn as VNode)
+      }
+  })
+
+  const movedNodes: movedNodesData[] = []
+  moveNodesSet.forEach(node => {
+      const nodeId = nodeStore.getNodeId(node)!
+      movedNodes.push({
+          parentId: nodeStore.getNodeId(node.parentNode!)!,
+          nextId: nodeStore.getNodeId(node.nextSibling!) || null,
+          id: nodeId
+      })
+  })
+
+  const removedNodes: RemoveUpdateData[] = []
+  removeNodesMap.forEach((parent, node) => {
+      const id = nodeStore.getNodeId(node)!
+      const parentId = nodeStore.getNodeId(parent)!
+
+      if (parentId) {
+          removedNodes.push({
+              parentId,
+              id
+          })
+      }
+  })
+
+  const attrs: AttributesUpdateData[] = attrNodesArray
+      .map(data => {
+          const { node, key, oldValue } = data
+          if (isExistingNode(node as Element)) {
+              const value = (node as Element).getAttribute(key)
+              if (oldValue === value) {
+                  return null
+              }
+              const id = nodeStore.getNodeId(node)
+              return {
+                  id,
+                  key,
+                  value
+              } as AttributesUpdateData
+          }
+      })
+      .filter(Boolean) as AttributesUpdateData[]
+
+  const texts = [...textNodesSet]
+      .map(textNode => {
+          if (isExistingNode(textNode) && textNode.parentNode) {
+              return {
+                  id: nodeStore.getNodeId(textNode),
+                  parentId: nodeStore.getNodeId(textNode.parentNode as Element),
+                  value: textNode.textContent
+              } as CharacterDataUpdateData
+          }
+      })
+      .filter(Boolean) as CharacterDataUpdateData[]
+
+  const data = {
+      addedNodes,
+      movedNodes,
+      removedNodes,
+      attrs,
+      texts
+  } as DOMUpdateDataType
+
+  if (Object.values(data).some(item => item.length)) {
+      emitterHook(emit, {
+          type: RecordType.DOM_UPDATE,
+          data,
+          time: getTime().toString()
+      })
+  }
+}
+
+export function DOMWatcher(options: WatcherOptions<DOMRecord>) {
+  const { emit, context } = options
+  const Watcher = new MutationObserver(callback => mutationCallback(callback, emit))
+
+  Watcher.observe(context.document.documentElement, {
+      attributeOldValue: true,
+      attributes: true,
+      characterData: true,
+      characterDataOldValue: true,
+      childList: true,
+      subtree: true
+  })
+
+  listenerStore.add(() => Watcher.disconnect())
+}
```

Comment:
add a new line to the end of a file
see: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

Reply:
i got it